Upgrade Storybook to 8.5

jandrade opened the pull request:
## Summary:

### Upgrading to v8.5.x

Upgrading Storybook to experiment with the new test addon.

We are doing this so we can test the new UI testing capabilities that work with
`vitest` + `Playwright` under the hood.

It will also get us closer to test the new a11y tooling that will be available
in the next days.

NOTE: This new feature is still under development by the Storybook team, so we
are not yet ready to use it in production.

You can see more information about this new testing approach here:

- [Storybook Test bootcamp](https://storybook.us18.list-manage.com/track/click?u=06a6fce3ab1327784d4342396&id=2a2ca202c1&e=50b9ba02d5)
- [Storybook Test addon docs](https://storybook.js.org/docs/writing-tests/test-addon)

### Integrating the new testing approach in CI

Now that we have upgraded to the experimental version, we can take advantage of its new features.

This PR adds a new step to our GH workflows: Running Storybook component tests in CI.

Issue: XXX-XXXX

## Test plan:

- Run `yarn storybook` and check if the new test addon is available.
- Also check that the Chromatic storybook environment still works.

Also verify how the storybook tests failed here:

https://github.com/Khan/wonder-blocks/actions/runs/12057096671/job/33620982112

<img width="1041" alt="Screenshot 2024-11-28 at 2 50 17 PM" src="https://github.com/user-attachments/assets/472fb90b-d98a-442f-9434-d55b3ae87f5f">

Now check that these tests were fixed here:

https://github.com/Khan/wonder-blocks/actions/runs/12074721182/job/33673231130#step:9:1024

<img width="1056" alt="Screenshot 2024-11-28 at 2 51 57 PM" src="https://github.com/user-attachments/assets/df2022d7-d64b-46f3-b740-bdfc052fce3b">

jandrade commented on __docs__/wonder-blocks-tooltip/tooltip.stories.tsx:
```diff
@@ -413,6 +396,23 @@ export const AutoUpdate: StoryComponentType = {
                 >
                     Click to update trigger position (fixed)
                 </Button>
+                <Tooltip
```
**note:** Moved this instance to the right so it can work more consistently in different environments (browser and JSDOM).

The issue with the previous story was that the assertion in the `play` function was not working properly in headless browsers (e.g. running `npx vitest`) because the viewport/document size was different from the browser one.

This was causing that the Tooltip was not positioned horizontally, causing the test to fail in CI. With this change, we can now ensure that there will be a horizontal reposition.

jandrade commented on vite.config.ts:
**note:** Moved this to a separate config file so it can be reused by `Storybook` and `vitest`.

jandrade commented on vitest.config.ts:
**note:** This file was generated when I added the experimental addon. More info here: https://storybook.js.org/docs/writing-tests/test-addon#automatic-setup

jandrade commented on .storybook/vitest.setup.ts:
**note:** This file was generated when I added the experimental addon. More info here: https://storybook.js.org/docs/writing-tests/test-addon#automatic-setup

jandrade commented on __docs__/wonder-blocks-link/link.stories.tsx:
```diff
@@ -83,10 +83,8 @@ Primary.play = async ({canvasElement}) => {
-    // compare different color formats, so hex was converted to RGB.
-    await expect(computedStyle.outline).toBe("rgb(24, 101, 242) solid 1px");
+    // rgb(24, 101, 242) is the same as Color.blue
+    await expect(link).toHaveStyle("outline: rgb(24, 101, 242) solid 1px");
```
**note:** This and other similar assertions where changed b/c `getComputedStyle` doesn't work with the new testing approach.

I got these errors the first time I integrated the new SB test CLI in CI, and made these changes to fix the failures.

beaesguerra replied:
What do we think about tests that check for specific styles? I had started removing jest tests that were asserting styles since it would be visually covered by Chromatic (especially with the "All Variants" stories + pseudostates add on!). Is it helpful to also have these component test checks for styles?  

Kinda related - In general, when would it be helpful for us to reach for component tests in Storybook? I'm thinking if there's functionality we want to test for that relies on browser behaviour! Curious to see what other think!

jandrade replied:
100% agree with you. We should move this to "All variants" but the issue is that `Link` (and `Button`) still rely on `Clickable` for setting pseudo states styles. As you know, I'd love to get us to a point where we don't use Clickable for those styles, and we let CSS handle that for us, but that's a tricky change that would be definitively nice to address.

Once we migrate to CSS pseudoclasses, I'm totally in favor of moving this to visual regression tests instead of interaction tests.

jandrade commented on __docs__/wonder-blocks-switch/switch.stories.tsx:
```diff
@@ -82,46 +82,50 @@ export const Default: StoryComponentType = {
-export const Controlled: StoryComponentType = () => {
-    const [checkedOne, setCheckedOne] = React.useState(false);
-    const [checkedTwo, setCheckedTwo] = React.useState(false);
+export const Controlled: StoryComponentType = {
```
**note:** This component doesn't have any big changes, just moved the code around to support CSF3.

jandrade commented:
Here's an example of how the new Test addon works locally:

https://github.com/user-attachments/assets/7541c41a-0aac-4a80-96a8-c2c8b9c8de6d

jeremywiebe commented on __docs__/wonder-blocks-dropdown/combobox.stories.tsx:
```diff
@@ -332,7 +332,7 @@ export const ControlledMultilpleCombobox: Story = {
 
     play: async ({canvasElement}) => {
-        const canvas = within(canvasElement);
+        const canvas = within(canvasElement.ownerDocument.body);
```
Interesting, this differs from the examples for these `play()` functions. :(  They always show how you originally had it.

jandrade replied:
Ahh yeah, this is useful for when we are using portals. I should add a comment here stating that.

jeremywiebe commented on package.json:
```diff
@@ -51,17 +52,16 @@
-    "@storybook/react": "^8.2.1",
-    "@storybook/react-vite": "^8.2.1",
-    "@storybook/test": "^8.2.1",
+    "@storybook/addon-a11y": "^8.5.0-alpha.10",
```
`-alpha.10` 😱   

Are you planning to wait till this is released before landing? 😁

jandrade replied:
hehe living on the edge! j/k

I can totally wait until it is officially released :). I mainly wanted to start this PR to be able to unlock their upcoming Accessibility addon (that uses this new testing tooling), so @marcysutton can try it out and give us some feedback, and also give feedback to the Storybook team as well, which she's planning to do :).

jeremywiebe reviewed (commented):
Very neat!  We'll have to see if component tests can be used in Perseus.

jandrade commented on .storybook/main.ts:
```diff
@@ -9,52 +6,17 @@ const config: StorybookConfig = {
         "../__docs__/**/*.mdx",
     ],
     addons: [
-        getAbsolutePath("@storybook/addon-essentials"),
```
oh I see, thanks for the context.

jandrade commented on .storybook/preview.tsx:
```diff
@@ -1,6 +1,6 @@
 import * as React from "react";
 import wonderBlocksTheme from "./wonder-blocks-theme";
-
+import {Decorator} from "@storybook/react";
```
Great point! I'll add this rule in a separate PR. I'll probably wait until WB is in a more "quiet" state to apply the changes :)

kevinb-khan commented:
> Here's an example of how the new Test addon works locally:

In the screen recording it looked like there were three errors, but when you clicked on each test individually it looked like they were all passing.  Am I misinterpreting what I was seeing?

kevinb-khan commented on .storybook/main.ts:
```diff
@@ -9,52 +6,17 @@ const config: StorybookConfig = {
         disableTelemetry: true,
     },
-    framework: getAbsolutePath("@storybook/react-vite"),
-    async viteFinal(config, {configType}) {
```
I see, it's just for running storybook tests.

jandrade replied:
correct! I'm thinking that we should probably write an ADR if we want to adopt this more widely.

Just curious.... I know this is a big change, but I wonder if you have any opinions about switching from Jest to vitest for unit tests?

kevinb-khan commented on __docs__/wonder-blocks-tooltip/tooltip.stories.tsx:
```diff
@@ -375,23 +375,6 @@ export const AutoUpdate: StoryComponentType = {
         } | null>(null);
         return (
             <View style={[styles.centered, styles.row, {position: "relative"}]}>
-                <Tooltip
-                    content="This is a tooltip that auto-updates its position when the trigger element changes."
-                    opened={true}
-                    autoUpdate={true}
-                >
-                    <View
-                        style={[
-                            position && {
-                                position: "absolute",
-                                top: position.y,
-                                left: position.x,
-                            },
-                        ]}
-                    >
-                        Trigger element
-                    </View>
-                </Tooltip>
```
Why was this removed?

jandrade replied:
It was moved to be included after the buttons. See my previous comment here: https://github.com/Khan/wonder-blocks/pull/2373/files#r1862635270

kevinb-khan commented on __docs__/wonder-blocks-switch/switch.stories.tsx:
```diff
@@ -82,46 +82,50 @@ export const Default: StoryComponentType = {
-    );
-};
-
-Controlled.play = async ({canvasElement}) => {
```
In previous versions of Storybook you had to manually navigate to each story and press "play" to run the tests, right?

jandrade replied:
Right! and that made it hard to determine when we introduced a regression with these interaction tests. I also added the CI step to ensure that we don't introduce issues in the future.

NOTE: Storybook uses the[ test-runner](https://storybook.js.org/docs/writing-tests/test-runner) for this, but it will be replaced by this new test addon.

The previous runner used Jest + Playwright, and they wanted to switch to Vitest specially because they can run these tests in browsers as well as in headless mode. This is a new `vitest` feature that is also being refined in collaboration between the Storybook and vitest teams: https://vitest.dev/guide/browser/

jandrade replied:
More context here: https://storybook.js.org/blog/storybook-test-sneak-peek/

kevinb-khan reviewed (approved):
LGTM.  You may want to wait until 8.5 is no longer alpha as Jeremy suggested.

jandrade commented:
> > Here's an example of how the new Test addon works locally:
> 
> In the screen recording it looked like there were three errors, but when you clicked on each test individually it looked like they were all passing. Am I misinterpreting what I was seeing?

@kevinb-khan good question. I added the video to use as a visual reference when we see failures in the browser. You can see the failures being reported in headless mode in CI here: https://github.com/Khan/wonder-blocks/actions/runs/12057096671/job/33620982112

What you see in the individual views, those tests were passing, but were considered flaky as it seemed that the results between browser and CI diverged. I asked the Storybook team about that, and they found the issue which I fixed here: https://github.com/Khan/wonder-blocks/pull/2373/commits/93279faf94d4c3315292d8bcd733478d75e240fc

The issues were not flaky per se, it was that I misconfigured the global decorators in preview.tsx.

For more context, the Storybook folks helped me to find the cause here: https://discord.com/channels/486522875931656193/1301551207835504694/1311423632412774481

I joined their #sb-test-early-access channel in the Storybook Discord server to be able to experiment with this new feature in our repo and get early feedback from them.

kevinb-khan commented:
> > > Here's an example of how the new Test addon works locally:
> > 
> > In the screen recording it looked like there were three errors, but when you clicked on each test individually it looked like they were all passing. Am I misinterpreting what I was seeing?
> 
> @kevinb-khan good question. I added the video to use as a visual reference when we see failures in the browser. You can see the failures being reported in headless mode in CI here: https://github.com/Khan/wonder-blocks/actions/runs/12057096671/job/33620982112

That's super cool that these tests run on CI in Chromatic.  This is really powerful stuff.  Thanks for setting this up.

beaesguerra reviewed (commented):
Thanks for setting this up! Looking forward to the upcoming a11y tooling 😄

jandrade commented on package.json:
```diff
@@ -110,17 +112,18 @@
     "typescript-coverage-report": "^0.7.0",
     "vite": "^4.4.8",
-    "vite-plugin-turbosnap": "^1.0.2"
+    "vitest": "^2.1.5"
```
**note:** This is a new dependency introduced in WB, but it's only required for the new Storybook testing approach. We are not going to use it directly at the moment.

jandrade commented on package.json:
```diff
@@ -110,17 +112,18 @@
     "jest-extended": "^4.0.2",
     "jscodeshift": "^0.15.1",
     "npm-package-json-lint": "^6.4.0",
+    "playwright": "^1.49.0",
```
**note:** This is a new dependency introduced in WB, but it's only required for the new Storybook testing approach. We are not going to use it directly at the moment.

jandrade commented:
@beaesguerra @jeremywiebe @kevinb-khan @marcysutton  Storybook 8.5 is now stable, I've updated this PR to use this version and everything works as expected: https://storybook.js.org/blog/storybook-8-5/

I'm going to land this change now! thanks for the reviews and feedback 🫶
